Docs content for Go drivers and ORMs (#3)
nchandrappa/yugabyte-db: docs/content/latest/drivers-orms/go/gorm.md

---
title: Go ORMs
linkTitle: Go ORMs
description: Go ORMs for YSQL
headcontent: Go ORMs for YSQL
image: /images/section_icons/sample-data/s_s1-sampledata-3x.png
menu:
  latest:
    name: Go ORMs
    identifier: gorm
    parent: go-drivers
    weight: 400
isTocNested: true
showAsideToc: true
---

<ul class="nav nav-tabs-alt nav-tabs-yb">

  <li >
    <a href="/latest/drivers-orms/go/gorm/" class="nav-link active">
      <i class="icon-postgres" aria-hidden="true"></i>
      GORM ORM
    </a>
  </li>

  <li >
    <a href="/latest/drivers-orms/go/pg/" class="nav-link">
      <i class="icon-postgres" aria-hidden="true"></i>
      PG ORM
    </a>
  </li>

</ul>

[GORM](https://gorm.io/) is the ORM library for Golang.

## Quick Start

Learn how to establish a connection to YugabyteDB database and begin simple CRUD operations using
the steps on the [Build an application](/latest/quick-start/build-apps/go/ysql-gorm) page under the
Quick start section.

Let us break down the quick start example and understand how to perform the common tasks required

> for Go App development using the GORM ORM tool.

Reviewer: Let's just call it "the GORM tool", or even just "using GORM." :)
```suggestion
for Go App development using GORM.
```

## Step 1: Import the Driver Package

You can import the GORM packages by adding the following import statement in your Go code.

> ### Import Statement
> 

Reviewer: These headings aren't required.
```suggestion
```

used for reading the data and writing the data into the database.

Code snippet for connecting to YugabyteDB:

```golang
 conn := fmt.Sprintf("host= %s port = %d user = %s password = %s dbname = %s sslmode=disable", host, port, user, password, dbname)
 var err error
 db, err = gorm.Open("postgres", conn)
 defer db.Close()
 if err != nil {
   panic(err)
 }
```

| Params | Description | Default |
| :---------- | :---------- | :------ |
| host  | hostname of the YugabyteDB instance | localhost
| port |  Listen port for YSQL | 5433
| user | user for connecting to the database | yugabyte
| password | password for connecting to the database | yugabyte
| dbname | database name | yugabyte

## Step 3: Create Table

Define a struct which maps to the table schema and use `AutoMigrate()` to create the table.

```golang
type Employee struct {
 Id       int64  `gorm:"primary_key"`
 Name     string `gorm:"size:255"`
 Age      int64
 Language string `gorm:"size:255"`
}
 ...

 // Create table
 db.Debug().AutoMigrate(&Employee{})
```

Read more on designing [Database schemas and tables](../../../../explore/ysql-language-features/databases-schemas-tables/).

## Step 4: Read and Write Data

> ### Insert Data
> 
> In order to write data into YugabyteDB, use the `db.Create()` functions.

Reviewer: ```suggestion
To write data into YugabyteDB, use the `db.Create()` function.
```

> ```golang

Reviewer: For all the snippets, tag them as "go" rather than "golang", just to be sure the highlighter catches them.
````suggestion
```go
````

Author: Okay, will do. It wasn't showing anything different for me.

 // Insert value
 db.Create(&Employee{Id: 1, Name: "John", Age: 35, Language: "Golang-GORM"})
 db.Create(&Employee{Id: 2, Name: "Smith", Age: 24, Language: "Golang-GORM"})
```

> ### Query Data
> 
> In order to query data from YugabyteDB tables, execute the `db.Find()` function.

Reviewer: ```suggestion
To query data from YugabyteDB tables, use the `db.Find()` function.
```
